Source/DafnyCore/Dafny.atg

```diff
@@ -3718,43 +3765,23 @@ LambdaSpec<.ref List<FrameExpression> reads, ref Attributes readsAttrs, ref Expr
 
 /*------------------------------------------------------------------------*/
 ParensExpression<out Expression e>
-= (. IToken lp; IToken rp;
+= (. IToken lp = null; IToken rp = null;
      var args = new List<ActualBinding>();
+     e = dummyExpr;
   .)
-  "("                                        (. lp = t; .)
-  [ TupleArgs<args> ]
-  ")"                                        (. rp = t; .)
-  (. if (args.Count == 1 && !args[0].IsGhost) {
-       if (args[0].FormalParameterName != null) {
-         SemErr(ErrorId.p_no_parenthesized_binding, new RangeToken(lp,rp), "binding not allowed in parenthesized expression");
-       }
-       e = new ParensExpression(lp, args[0].Actual);
-     } else {
-       // Compute the actual position of ghost arguments
-       var ghostness = new bool[args.Count];
-       for (var i = 0; i < args.Count; i++) {
-         ghostness[i] = false;
-       }
-       for (var i = 0; i < args.Count; i++) {
-         var arg = args[i];
-         if (arg.IsGhost) {
-           if (arg.FormalParameterName == null) {
-             ghostness[i] = true;
-           } else {
-             var success = int.TryParse(arg.FormalParameterName.val, out var index);
-             if (success && 0 <= index && index < args.Count) {
-               ghostness[index] = true;
-             }
-           }
-         }
-       }
-       var argumentGhostness = ghostness.ToList();
-       // make sure the corresponding tuple type exists
-       SystemModuleModifiers.Add(b => b.TupleType(lp, args.Count, true, argumentGhostness));
-       e = new DatatypeValue(lp, SystemModuleManager.TupleTypeName(argumentGhostness), SystemModuleManager.TupleTypeCtorName(args.Count), args);
-     }
-     e.RangeToken = new RangeToken(lp, rp);
-  .)
+ "("                                        (. lp = t; .)
+  (
+    MaybeDecreasesToExpression<out e, lp>
+    ")"                                     (. rp = t;
+                                               if (e is not DatatypeValue) {
```

Comment:
Is a `DecreasesToExpr` always wrapped in a `ParensExpression` ? That seems unnecessary. Considering adding `&& e is not DecreasesToExpr`

Also, what if you write `(Nil)`, does this then not get wrapped in a `ParensExpression` ?

The previous check, `args.Count == 1 && !args[0].IsGhost`, to me seems to more accurately capture that only if you parsed a tuple body with multiple elements or a ghost singleton tuple, may you omit the ParensExpression.